src/script/taproot.cpp

```diff
@@ -62,3 +63,24 @@ bool VerifyTaprootCommitment(uint256 &tapleaf_hash,
     }
     return q == q_expected;
 }
+
+bool IsPayToTaproot(const CScript &script) {
+    if (script.size() < TAPROOT_SIZE_WITHOUT_STATE) {
+        return false;
+    }
+    // Taproot must start with OP_SCRIPTTYPE OP_1
+    if (script[0] != OP_SCRIPTTYPE || script[1] != TAPROOT_SCRIPTTYPE) {
+        return false;
+    }
+    // First push (commitment) must be 33 bytes long
+    if (script[2] != CPubKey::COMPRESSED_SIZE) {
+        return false;
+    }
+    // If there's only a commitment but no state, we're valid
+    if (script.size() == TAPROOT_SIZE_WITHOUT_STATE) {
+        return true;
+    }
+    // Otherwise, we need a state with 32 bytes.
+    return script.size() == TAPROOT_SIZE_WITH_STATE &&
```

Comment:
Are we going to be allowing state in the first iteration?

Reply:
I’d say so, otherwise the smart contract capabilities are severely limited. We’d need EC_MUL and EC_ADD to do recursive smart contracts